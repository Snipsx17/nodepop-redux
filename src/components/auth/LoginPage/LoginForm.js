import T from 'prop-types';
import { Form, FormConsumer, Input } from '../../form';

const validEmail = ({ email }) => email;
const validPassword = ({ password }) => password;

function LoginForm({ onSubmit, isLoading }) {
  return (
    <Form
      onSubmit={onSubmit}
      initialValue={{
        email: '',
        password: '',
        remember: false,
      }}
    >
      <label>
        email: <Input name="email" />
      </label>
      <label>
        password:
        <Input type="password" name="password" />
      </label>
      <Input type="checkbox" name="remember" />
      <FormConsumer>
        {({ validate }) => (
          <button
            disabled={!validate(validEmail, validPassword, () => !isLoading)}
          >
            Login
          </button>
        )}
      </FormConsumer>
    </Form>
  );
}

LoginForm.propTypes = {
  isLoading: T.bool,
  onSubmit: T.func.isRequired,
};

export default LoginForm;
