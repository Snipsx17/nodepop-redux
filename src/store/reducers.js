import {
  ADVERTS_CREATED_SUCCESS,
  ADVERTS_DELETED_SUCCESS,
  ADVERTS_DETAIL_SUCCESS,
  ADVERTS_LOADED_SUCCESS,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  TAGS_LOADED,
  UI_RESET_ERROR,
} from './types';

export const initState = {
  auth: false,
  adverts: {
    loaded: false,
    data: [],
  },
  tags: [],
  ui: {
    loading: false,
    error: null,
  },
};
//AUTH
export function auth(state = initState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT_SUCCESS:
      return false;
    default:
      return state;
  }
}
//ADVERTS
export const adverts = (state = initState.adverts, action) => {
  switch (action.type) {
    case ADVERTS_LOADED_SUCCESS:
      return { loaded: true, data: action.payload };
    case ADVERTS_DETAIL_SUCCESS:
      return { ...state, data: [action.payload] };

    case ADVERTS_CREATED_SUCCESS:
      return { ...state, data: [...state.data, action.payload] };
    case ADVERTS_DELETED_SUCCESS:
      return {
        ...state,
        data: state.data.filter((advert) => advert.id !== action.payload),
      };
    default:
      return state;
  }
};
//TAGS
export const tags = (state = initState.tags, action) =>
  action.type === TAGS_LOADED ? action.payload : state;
//UI
export function ui(state = initState.ui, action) {
  if (action.error) {
    return { loading: false, error: action.payload };
  }
  if (action.type.endsWith('/request')) {
    return { loading: true, error: null };
  }
  if (action.type.endsWith('/success')) {
    return { loading: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }
  return state;
}
