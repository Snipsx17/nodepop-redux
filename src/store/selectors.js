//ADVERTS
export const getAreAdvertsLoaded = (state) => state.adverts.loaded;
export const getAdverts = (state) => state.adverts.data;
export const getAdvert = (advertId) => (state) =>
  getAdverts(state).find((advert) => advert.id === advertId);
//AUTH
export const getIsLogged = (state) => state.auth;
//UI
export const getUi = (state) => state.ui;
//TAGS
export const getTags = (state) => state.tags;
export const getAreTagsLoaded = (state) => getTags(state).length > 0;
