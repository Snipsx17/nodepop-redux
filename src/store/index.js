import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import { withExtraArgument } from 'redux-thunk';
import * as reducers from './reducers';
import { adverts, auth } from '../service';
import * as actionCreators from './actions';

const composeEnhancers = composeWithDevTools({ actionCreators });

const failureRedirections =
  (router, redirectMap) => (store) => (next) => (action) => {
    const result = next(action);
    if (action.error) {
      const to = redirectMap[action.payload.statusCode];
      if (to) {
        router.navigate(to);
      }
    }
    return result;
  };

export default function configureStore(preloadedState, { router }) {
  const middlewares = [
    withExtraArgument({ api: { adverts, auth }, router }),
    failureRedirections(router, {
      401: '/login',
      404: '/404',
    }),
  ];

  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    composeEnhancers(applyMiddleware(...middlewares))
  );

  return store;
}
