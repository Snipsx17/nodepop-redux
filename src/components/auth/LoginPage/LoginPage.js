import React from 'react';

import LoginForm from './LoginForm';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { getUi } from '../../../store/selectors';

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const handleSubmit = credentials => {
    dispatch(authLogin(credentials));
  };

  const handleErrorClick = () => {
    dispatch(uiResetError());
  };

  return (
    <div>
      <LoginForm onSubmit={handleSubmit} isLoading={isLoading} />
      {isLoading && <p>...login in nodepop</p>}
      {error && (
        <div onClick={handleErrorClick} style={{ color: 'red' }}>
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
