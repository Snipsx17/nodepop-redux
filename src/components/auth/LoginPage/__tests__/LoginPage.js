import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { authLogin } from '../../../../store/actions';
import { initState } from '../../../../store/reducers';
import LoginPage from '../LoginPage';
import userEvent from '@testing-library/user-event';

jest.mock('../../../../store/actions');

describe('LoginPage', () => {
  const store = {
    getState: () => initState,
    dispatch: () => {},
    subscribe: () => {},
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

  //Create snapshot
  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should call authLogin', async () => {
    const email = 'email@email.com';
    const password = '1234';

    renderComponent();

    const inputEmail = screen.getByLabelText(/email/);
    const inputPassword = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();
    await userEvent.type(inputEmail, email);
    await userEvent.type(inputPassword, password);
    await userEvent.click(submitButton);

    fireEvent.click(submitButton);
    expect(authLogin).toHaveBeenCalledWith({
      email,
      password,
      remember: false,
    });
  });
});
