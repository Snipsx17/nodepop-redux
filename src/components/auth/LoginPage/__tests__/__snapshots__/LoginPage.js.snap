// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div>
    <form>
      <label>
        email: 
        <input
          name="email"
          value=""
        />
      </label>
      <label>
        password:
        <input
          name="password"
          type="password"
          value=""
        />
      </label>
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
